### DISCUSSION TITLE
OPA development documentation

### DISCUSSION THREAD

**Discussion by niuzhi** [open-policy-agent/community#167] [OPA and Rego]
Hello，I am reading the source code of OPA, may I ask if there are development documents or materials for OPA so that I can master OPA and start contributing code to the community, thank you!

**Comment by anderseknert** [comment]
Hi @niuzhi ! And welcome to the OPA contributor community 😃

An introduction to OPA development is found in the [OPA docs](https://www.openpolicyagent.org/docs/edge/contrib-development/), and front-end/docs development is described [here](https://github.com/open-policy-agent/opa/tree/main/docs).

If you'd like to get started on some issue, the [good-first-issue](https://github.com/open-policy-agent/opa/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) label could be a good starting point.

Additionally, the #development channel on the OPA Slack is a great place to ask questions, or just follow along in what others might be working with.

Or just ask away here. Let me know if you have any questions!

**Comment by niuzhi** [comment]
@anderseknert Thanks，I am reading the code of the eval command in opa, but I found that it is a bit confusing after reading evalOnce. Can you explain the business process related to the output of the query result after evalOnce, so that I can read the code. If you can help me solve it, it will help me Understanding OPA helps a lot, thanks a lot!
![image](https://user-images.githubusercontent.com/33443724/163084974-382e7d90-c7d0-4c40-8e7c-76692713b52f.png)

**Reply by anderseknert** [reply]
The `eval` command, and by extension the functions that it calls, just takes some input and evaluates that against some given policy and data, and eventually returns the result of the evaluation. Is there any specific part of that code which you'd want me to explain?

**Comment by niuzhi** [comment]
@anderseknert Thank you for your reply. If it doesn't affect your time, could you please briefly explain his query process. At present, this code is a bit confusing, that is, err := e.Run(func(e *eval) returns the result at the end The relevant rationale for the results. Many thanks
![image](https://user-images.githubusercontent.com/33443724/163178884-673678d0-bb1f-4b1b-8493-9376bc95bb9e.png)

**Reply by anderseknert** [reply]
Right, so the outcome of an `opa eval` invocation could not just a single value, but a set of bindings. You'll see this if you run something like this:

```shell
❯ opa eval 'x := 5; y := 10'
{
  "result": [
    {
      "expressions": [
        {
          "value": true,
          "text": "x := 5",
          "location": {
            "row": 1,
            "col": 1
          }
        },
        {
          "value": true,
          "text": "y := 10",
          "location": {
            "row": 1,
            "col": 9
          }
        }
      ],
      "bindings": {
        "x": 5,
        "y": 10
      }
    }
  ]
}
```
